File: year2/week22/solution/week22.hs
majorityCandidate :: (Eq a) => [a] -> a
majorityCandidate (x:xs) = fst $ foldr majorityCandidate' (x, 1) xs
    where
        majorityCandidate' y (x, c)
            | c == 0    = (y, 1)
            | x == y    = (x, c + 1)
            | otherwise = (x, c - 1)

isMajority :: (Eq a) => a -> [a] -> Bool
isMajority x xs = count x xs > length xs `div` 2
    where count x = length . filter (==x)

majorityElement :: (Eq a) => [a] -> Maybe a
majorityElement xs
    | isMajority x xs = Just x
    | otherwise       = Nothing
        where x = majorityCandidate xs

main = lines <$> getContents >>= output . majorityElement
    where
        output (Just x) = putStrLn x
        output Nothing  = putStrLn "no majority"

File: year2/week22/solution/week22.py
import sys

def majority_candidate(votes):
    def count((candidate, count), name):
        if count == 0:
            return name, 1
        if name == candidate:
            return candidate, count + 1
        return candidate, count - 1

    candidate, _ = reduce(count, votes, (votes[0], 0))
    return candidate

def is_majority(value, iterable):
    return iterable.count(value) > len(iterable) / 2

def majority_element(iterable):
    x = majority_candidate(iterable)
    if is_majority(x, iterable):
        return x
    return None

def main():
    votes = sys.stdin.readlines()
    prez = majority_element(votes)
    print prez or "no majority"

if __name__ == "__main__":
    main()
